File: src/RayTracing/TypeScript/Sphere.ts
import Ray from "./Ray";
import { SolveQuadratic } from "./Solver";
import Vec, { Vector3 } from "./Vector3";

class Sphere {
  origin: Vector3;
  radius: number;

  constructor(origin: Vector3, radius: number) {
    this.origin = origin;
    this.radius = radius;
  }

  intersect(ray: Ray): number | null {
    // working out in whiteboard
    const newOrigin = Vec.sub(ray.origin, this.origin);

    const a: number = 1;
    const b: number = 2 * Vec.dot(ray.direction, newOrigin);
    const c: number = Vec.dot(newOrigin, newOrigin) - (this.radius ** 2);

    const solutions = SolveQuadratic(a, b, c);

    // doesn't intersect
    if (solutions === null) return null;

    // just touches at 1 point
    if (!Array.isArray(solutions)) return solutions;

    // return the minimum
    return solutions[0] < solutions[1] ? solutions[0] : solutions[1];
  }

  normalAtPoint(point: Vector3): Vector3 {
    // simple circle stuff
    const normal = Vec.normalize(Vec.sub(point, this.origin));
    return normal;
  }
}

export default Sphere;

File: src/RayTracing/rs-ray-tracing/src/sphere.rs
use crate::vec3::Vec3;
use crate::ray::Ray;
pub use crate::solver::solve_quadratic;

#[derive(Clone, Copy)]
pub struct Sphere {
  pub center: Vec3,
  pub radius: f64,
}

impl Sphere {
  pub fn intersect (&self, ray: &Ray) -> Option<f64> {
    // working out in whiteboard
    let new_origin = ray.origin - self.center;

    let a = 1.;
    let b = 2. * ray.direction.dot(new_origin);
    let c = new_origin.dot(new_origin) - self.radius.powi(2);

    let solution = solve_quadratic(a, b, c);

    match solution {
        Some(solution) => {
          if solution.0 < solution.1 {
            Some(solution.0)
          } else {
            Some(solution.1)
          }
        }
        None => None
    }
  }

  pub fn normal_at_point (&self, point: Vec3) -> Vec3 {
    // simple circle stuff
    (point - self.center).normalize()
  }
}
